src/box/txn_limbo.h

```diff
@@ -266,6 +276,59 @@ txn_limbo_replica_term(const struct txn_limbo *limbo, uint32_t replica_id)
 	return vclock_get(&limbo->promote_term_map, replica_id);
 }
 
+/**
+ * Return true if the promote data is locked.
+ */
+static inline bool
+txn_limbo_promote_is_locked(const struct txn_limbo *limbo)
+{
+	return limbo->promote_lock.owner != NULL;
+}
+
+/**
+ * Lock the promote data.
+ * If it is already locked by another fiber, wait for it.
+ * Locks are acquired in the strict order as they were requested.
+ */
+static inline void
+txn_limbo_promote_lock(struct txn_limbo *limbo,
+		       struct txn_limbo_promote_lock *waiter)
+{
+	assert(limbo->promote_lock.owner != fiber());
+	struct rlist *queue = &limbo->promote_lock.queue;
+	if (limbo->promote_lock.owner == NULL && rlist_empty(queue)) {
+		limbo->promote_lock.owner = fiber();
+		return;
+	}
+
+	waiter->owner = fiber();
+	rlist_add_tail_entry(queue, waiter, queue);
+	do {
+		fiber_yield();
+	} while (limbo->promote_lock.owner != NULL);
+
+	limbo->promote_lock.owner = fiber();
+	rlist_del_entry(waiter, queue);
```

Comment:
Looks like this code could be reused somewhere else. Maybe, add a generic data structure for this ordered latch?

Reply:
> Looks like this code could be reused somewhere else. Maybe, add a generic data structure for this ordered latch?

+1

Reply:
Done.